lang/java/avro/src/test/java/org/apache/avro/TestSchemaValidation.java

```diff
@@ -15,183 +15,316 @@
+
+      new ReaderWriter(ENUM1_AB_SCHEMA, ENUM1_ABC_SCHEMA),
+      new ReaderWriter(ENUM1_BC_SCHEMA, ENUM1_ABC_SCHEMA),
+
+      new ReaderWriter(ENUM1_AB_SCHEMA, ENUM2_AB_SCHEMA),
+      new ReaderWriter(INT_SCHEMA, ENUM2_AB_SCHEMA),
+      new ReaderWriter(ENUM2_AB_SCHEMA, INT_SCHEMA),
+
+      // Tests involving unions:
+      new ReaderWriter(INT_UNION_SCHEMA, INT_STRING_UNION_SCHEMA),
+      new ReaderWriter(STRING_UNION_SCHEMA, INT_STRING_UNION_SCHEMA),
+      new ReaderWriter(FLOAT_SCHEMA, INT_LONG_FLOAT_DOUBLE_UNION_SCHEMA),
+      new ReaderWriter(LONG_SCHEMA, INT_FLOAT_UNION_SCHEMA),
+      new ReaderWriter(INT_SCHEMA, INT_FLOAT_UNION_SCHEMA),
+
+      new ReaderWriter(EMPTY_RECORD2, EMPTY_RECORD1),
+      new ReaderWriter(A_INT_RECORD1, EMPTY_RECORD1),
+      new ReaderWriter(A_INT_B_DINT_RECORD1, EMPTY_RECORD1),
+
+      new ReaderWriter(INT_LIST_RECORD, LONG_LIST_RECORD),
+
+      // Last check:
+      new ReaderWriter(NULL_SCHEMA, INT_SCHEMA));
+
+  SchemaValidatorBuilder builder = new SchemaValidatorBuilder();
+  Schema rec = SchemaBuilder.record("test.Rec").fields().name("a").type().intType().intDefault(1).name("b").type()
+      .longType().noDefault().endRecord();
+  Schema rec2 = SchemaBuilder.record("test.Rec").fields().name("a").type().intType().intDefault(1).name("b").type()
+      .longType().noDefault().name("c").type().intType().intDefault(0).endRecord();
+  Schema rec3 = SchemaBuilder.record("test.Rec").fields().name("b").type().longType().noDefault().name("c").type()
+      .intType().intDefault(0).endRecord();
+  Schema rec4 = SchemaBuilder.record("test.Rec").fields().name("b").type().longType().noDefault().name("c").type()
+      .intType().noDefault().endRecord();
+  Schema rec5 = SchemaBuilder.record("test.Rec").fields().name("a").type().stringType().stringDefault("") // different
+                                                                                                          // type
+                                                                                                          // from
+                                                                                                          // original
+      .name("b").type().longType().noDefault().name("c").type().intType().intDefault(0).endRecord();
+  Schema s = SchemaBuilder.record("r").fields().requiredBoolean("boolF").requiredInt("intF").requiredLong("longF")
+      .requiredFloat("floatF").requiredDouble("doubleF").requiredString("stringF").requiredBytes("bytesF").name(
+          "fixedF1")
+      .type().fixed("F1").size(1).noDefault().name("enumF").type().enumeration("E1").symbols("S").noDefault()
+      .name("mapF").type().map().values().stringType().noDefault().name(
+          "arrayF")
+      .type().array().items().stringType().noDefault().name("recordF").type().record("inner").fields().name("f").type()
+      .intType().noDefault().endRecord().noDefault().optionalBoolean("boolO").endRecord();
 
   public static class Point {
```

Comment:
What happened to testReflectMatchStructure and testReflectWithAllowNullMatchStructure test cases? Why are they gone? If there's no need for them, then I think you should also get rid of Point and Circle classes as well as circleSchemaDifferentNames and circleSchema.